### PR TITLE
Support syncing global storage

### PR DESCRIPTION
**Pull request by sandy081**
This PR implements https://github.com/microsoft/vscode/issues/86679

We store UI state in our global storage and to sync UI state we have to sync global storage. 
But we cannot sync all global storage because it can have unnecessary data and also some of that state is scoped to machine. Hence sync needs to know what data in global storage has to be synced. 

Also it happens (not often but can) that we change the semantics of data in global state and handle migration (by the corresponding data owners). Since there is a greater possibility that VS Code instances with different versions can be syncing data we have to make sure old data does not enter into the system. 

With above two requirements, I have introduced storage keys sync registry. Storage data owners can register their data key with current version. 

```ts
export interface IStorageKey {

	readonly key: string;
	readonly version: number;

}
```

With this information synchroniser will sync only data of registered storage keys. Also checks version of incoming data and make sure it is not entered into the system. Also true while updating the data in the server that old data is not updated into the server.

@bpasero 
Creating the PR with above design. Please let me know your feedback.
Work is still in progress.

### CODE REVIEW

File: src/vs/workbench/browser/parts/activitybar/activitybarPart.ts [hunk 1]
```diff
@@ -92,7 +92,7 @@ export class ActivitybarPart extends Part implements IActivityBarService {
 		@IContextKeyService private readonly contextKeyService: IContextKeyService,
 		@IConfigurationService private readonly configurationService: IConfigurationService,
 		@IWorkbenchEnvironmentService workbenchEnvironmentService: IWorkbenchEnvironmentService,
-		@IEnvironmentService private readonly environmentService: IEnvironmentService
+		@IEnvironmentService private readonly environmentService: IEnvironmentService,
```
Comment (bpasero) [inline src/vs/workbench/browser/parts/activitybar/activitybarPart.ts]:
```suggestion
		@IEnvironmentService private readonly environmentService: IEnvironmentService
```

**Review by bpasero — APPROVED** [review]
This seems legit looking at the code, though arguably a lot of infrastructure for knowing which keys to sync (maybe if there is already some sync service we could piggy-bag on to not introduce a new service?).

But raises some questions:
* how can an extension signal that its global state should be synced if we do this filtering?
* should the user be able to configure this instead of us hardcoding it?
* how does this work in the web?

**Comment by sandy081** [comment]
@bpasero  Thanks for suggestions.

There is a global `IUserDataSyncService`. But having this functionality in this service will cause cyclic dep. I like to have this separate.

> how can an extension signal that its global state should be synced if we do this filtering?

Thinking of introducing a new contribution point for allowing extensions to register global state to be synced.

> should the user be able to configure this instead of us hardcoding it?

Thought about it - like configuring ignored settings or extensions. But did not go with that because global state is not configurable. I would push back this request until it is highly requested.

> how does this work in the web?

Similar to Desktop client. Global state sync also runs in web and updates the storage by registered sync keys.

**Comment by bpasero** [comment]
@sandy081 maybe make this service more general for what to include or exclude in sync overall? It seems to me that we already have a concept of excluding something in settings if the user decides so. We possibly also have some default setting keys we decide to not sync. So this service could:
* for each area to sync
* allow to opt-in or opt-out of syncing programmatically
* detect for some areas by looking into the file itself (e.g. settings)

**Comment by sandy081** [comment]
Sure. I already have one such service `IUserDataSyncUtliService` (not a good name but could not find a better one)  that's meant to have some area (settings, keybindings) specific util methods. 
I can move it there.

**Comment by bpasero** [comment]
@sandy081 yeah `resolveDefaultIgnoredSettings` looks very related on the settings side

**Comment by sandy081** [comment]
@bpasero But there is a problem that this service lives on renderer. The storage keys related service shall be in `main` as we might also have storage keys on main which might be needed to sync. So this shall be similar to storage service. May be better is to leave it as it is (separate service)?

**Comment by bpasero** [comment]
@sandy081 could it not live in `platform` and be accessed from any process? What makes it renderer specific?

**Comment by bpasero** [comment]
Btw I am not feeling strong about it if it is a lot of work and no real gain...

**Comment by sandy081** [comment]
Some functions in `UserDataSyncUtilService` can only be implemented by renderer. For eg: `getDefaultIgnoredSettings` has to read all configurations and the registry is available only in renderer.

**Comment by sandy081** [comment]
After looking at other available services.. I feel it is good to be a separate service.

**Pull request merged**